### ISSUE TITLE
training: shuffle data between epochs

### ISSUE THREAD

**Issue by wosiu** [Calamari-OCR/calamari#74]
First of all - thank you for that fantastic framework! I've been using tesseract for more than 1 year, but this one is way better for a single line processing :)

Proposal:
From the logs during training, it seems that input images are not shuffled at all.
It would be nice, if they are shuffled at least at the very beginning.
And it would be perfect if data are also shuffled after each epoch, so that different batches are created.

**Comment by wosiu** [comment]
I could try to investigate that and make PR. Let me know if that makes sense to you. Any hints are appreciated.

**Comment by ChWick** [comment]
See [here](https://github.com/Calamari-OCR/calamari/blob/8a74f401fe1e5b83c5783752112023c9ba133a50/calamari_ocr/ocr/backends/tensorflow_backend/tensorflow_model.py#L251):
```
dataset = dataset.repeat().shuffle(buffer_size, seed=self.network_proto.backend.random_seed)
```

The dataset is shuffled using the tensorflow  api (not during loading), however only up to a maximum of `buffer_size` (by default 1000 instances).

**Comment by wosiu** [comment]
What is the reason to have this `buffer_size` limit for deciding whether to shuffle data or not?
Why 1000 by default?

**Comment by ChWick** [comment]
The reason is, that this size is direcyly related to the required RAM but there is no deeper reason for this size. We chose this value, because it ensures that Calamari runs also on weaker/older laptops that are in use.
Probably, one should make this a parameter and changeably via the command line (or network parameters).

**Comment by wosiu** [comment]
+1 I'll try then to develop it and PR in free time. Please keep this issue open.

**Comment by ChWick** [comment]
@wosiu in 7b3be26 I added a training parameter to define the buffer size (--shuffle_buffer_size). Use 0 if you want the shuffle buffer the same size as the dataset, i. e. one epoch.

**Comment by wosiu** [comment]
Looks perfect, many thanks :) Closing.
